File: src/cpp/poller.cpp
#include "poller.h"

#include <zmq.hpp>
#include <stdexcept>

#include "error.h"

using namespace bjcomm;

Poller::Poller(): _items(nullptr) {}
Poller::~Poller() {
    delete[] _items;
}

int Poller::add(CommunicationInterface *interface){
    if(!interface->isRunning()) return -1;
    _interfaces.push_back(interface);
    return _interfaces.size()-1;
}

//TODO: make the implementation faster (we dont use more than 2 sockets most probably though)
void Poller::poll(){
    delete[] _items;
    
    //check if interfaces are added
    if(_interfaces.empty()) throw CommunicationError("Polling without any interfaces added!");
    
    _items = new zmq::pollitem_t[_interfaces.size()];
    for(size_t i=0; i<_interfaces.size(); ++i){
        //return if some interface is not running (this causes busy waiting...)
        if(!_interfaces[i]->isRunning()){
            delete[] _items;
            throw CommunicationError("Polling on a channel that is not running!");
        }
        _items[i] = {(void *) *(_interfaces[i]->_socket), 0, ZMQ_POLLIN, 0 };
    }

    zmq::poll (_items, _interfaces.size(), -1);
}

bool Poller::hasMsg(int id){
    //if no poll or invalid id throw error
    if(id < 0 || id >= (int) _interfaces.size()) throw std::invalid_argument("Checking for message of an invalid identifier!");
    else if(_items == nullptr) throw std::logic_error("Checking for a message before poller is called!");
    
    return _items[id].revents & ZMQ_POLLIN;
    
}

File: src/java/Poller.java
/*
 * Copyright (c) 2015 Blue Jay Eindhoven
 */
package nl.bluejayeindhoven.bjcomm;

import nl.bluejayeindhoven.bjcomm.CommunicationInterface;

import org.zeromq.ZMQ;
import org.zeromq.ZMQException;

/**
 * Implement simple poller that divides load between several channels
 *
 * @author Koen Wolters
 */
public class Poller{
    private ZMQ.Poller poller;
    boolean polled;
    
    /**
     * Initializes a poller
     */
    public Poller(){
        poller = new ZMQ.Poller(0);
        polled = false;
    }
    
    /**
     * Adds an interface to the poller. Make sure the interface is loaded before adding it. Only add interfaces that support polling.
     * 
     * @param comm interface to add to the poller
     * @return identifier for the added interface or -1 if failed
     */
    public int add(CommunicationInterface comm){
        if(comm.socket == null) return -1;
        return poller.register(comm.socket, ZMQ.Poller.POLLIN);
    }
    
    /** 
     * Poll all registered interfaces. Blocks until at least one of them has a message to process.
     *
     * @throws CommunicationError if an added interface is not started
     */
    public void poll(){
        if(poller.getSize() == 0) throw new CommunicationError("Polling without any interfaces added!");
        
        polled = false;
        try{
            poller.poll(-1);
            polled = true;
        }catch(ZMQException zex){
            throw new CommunicationError("Polling on an channel that is not running!");
        }
    }
    
    /**
     * Check if the supplied interface has a message to read
     *
     * @param id identifier of an interface returned by {@link #add(CommunicationInterface) add} method.
     * @return if the interface has a message to read
     *
     * @throws IllegalStateException if called before {@link #poll() poll()} is called
     * @throws IllegalArgumentException if called with an invalid identifier
     */
    public boolean hasMsg(int id){
        //check if id exists and poll has been called
        if(id < 0 || id >= poller.getSize()) throw new IllegalArgumentException("Checking for message of an invalid identifier!");
        else if(!polled) throw new IllegalStateException("Checking for a message before poller is called!");
        
        return poller.pollin(id);
    }
}